independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/util/QuarkusModelHelper.java

```diff
@@ -0,0 +1,179 @@
+        final Path serializedModel = File.createTempFile("quarkus-app-model", ".dat").toPath();
+        final ArtifactCoords artifactCoords = model.getWorkspace().getMainModule().getArtifactCoords();
+        AppArtifact appArtifact = new AppArtifact(artifactCoords.getGroupId(),
+                artifactCoords.getArtifactId(),
+                artifactCoords.getVersion());
+        try (ObjectOutputStream out = new ObjectOutputStream(Files.newOutputStream(serializedModel))) {
+            out.writeObject(QuarkusModelHelper.convert(model, appArtifact));
+        }
+        return serializedModel;
+    }
+
+    public static Path getClassPath(WorkspaceModule model) throws BootstrapGradleException {
+        // TODO handle multiple class directory
+        final Optional<Path> classDir = model.getSourceSet().getSourceDirectories().stream().filter(File::exists)
+                .map(File::toPath).findFirst();
+        if (!classDir.isPresent()) {
+            throw new BootstrapGradleException("Failed to locate class directory");
+        }
+        return classDir.get();
+    }
+
+    public static AppModel convert(QuarkusModel model, AppArtifact appArtifact) throws AppModelResolverException {
+        AppModel.Builder appBuilder = new AppModel.Builder();
+
+        final List<AppDependency> userDeps = new ArrayList<>();
+        Map<AppArtifactKey, AppDependency> versionMap = new HashMap<>();
+        model.getAppDependencies().stream().map(QuarkusModelHelper::toAppDependency).forEach(appDependency -> {
+            userDeps.add(appDependency);
+            versionMap.put(appArtifact.getKey(), appDependency);
+        });
+
+        final List<AppDependency> deploymentDeps = new ArrayList<>();
+        for (Dependency extensionDependency : model.getExtensionDependencies()) {
+            AppDependency appDep = toAppDependency(extensionDependency);
+            for (Path artifactPath : appDep.getArtifact().getPaths()) {
+                if (!Files.exists(artifactPath) || !extensionDependency.getType().equals("jar")) {
+                    continue;
+                }
+                if (Files.isDirectory(artifactPath)) {
+                    processQuarkusDir(appDep.getArtifact(), artifactPath.resolve(BootstrapConstants.META_INF),
+                            appBuilder);
+                } else {
+                    try (FileSystem artifactFs = FileSystems.newFileSystem(artifactPath,
+                            QuarkusModelHelper.class.getClassLoader())) {
+                        processQuarkusDir(appDep.getArtifact(), artifactFs.getPath(BootstrapConstants.META_INF),
+                                appBuilder);
+                    } catch (IOException e) {
+                        throw new AppModelResolverException("Failed to process " + artifactPath, e);
+                    }
+                }
+            }
+            if (!userDeps.contains(appDep)) {
```

Comment:
This is not correct. You need to be comparing the keys instead. Otherwise, you are not aligning versions here.

Reply:
This is actually done like this in the current model, should i update it to use the `appDep.getArtifact()` instead? (https://github.com/quarkusio/quarkus/blob/23b3d64ead8d67a4ed6e5b88a4853f0a8aaefe20/devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java#L189)

Reply:
Oh, that's clearly a bug. It should be `appDep.getArtifact().getKey()`